sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/kinesis/enhancedfanout/EFOShardSubscriber.java

```diff
@@ -126,31 +139,26 @@ public EFOShardSubscriber(
       String shardId,
       KinesisIO.Read read,
       KinesisAsyncClient kinesis,
-      long onErrorCoolDownMs) {
+      int onErrorCoolDownMs) {
     this.pool = pool;
     this.consumerArn = checkArgumentNotNull(read.getConsumerArn());
     this.shardId = shardId;
     this.kinesis = kinesis;
-    this.inFlight = new AtomicInteger();
     this.reSubscriptionHandler =
         (Void unused, Throwable error) -> {
           eventsSubscriber.cancel();
           if (error != null && !isRetryAble(error)) {
             done.completeExceptionally(error);
-          } else if (!isStopped) {
+          } else if (state != STOPPED) {
             String lastContinuationSequenceNumber = eventsSubscriber.sequenceNumber;
+            // FIXME must resubscribe from initial starting position if retryable error && lastContinuationSequenceNumber == null
             if (lastContinuationSequenceNumber == null) {
               done.complete(null); // completely consumed this shard, done
+            } else if (error != null && inFlight.get() == IN_FLIGHT_LIMIT) {
+              state = PAUSED;
```

Comment:
This should help to avoid repeated re-subscriptions if the reader is stuck: We simply delay resubscribe until the next event is acked.